x-pack/plugins/security_solution/public/common/utils/validators/index.test.ts

```diff
@@ -13,6 +13,10 @@ describe('helpers', () => {
       expect(isUrlInvalid('this is not a url')).toBeTruthy();
     });
 
+    test('verifies as invalid url without http(s):// prefix', () => {
+      expect(isUrlInvalid('www.thisIsNotValid.com')).toBeTruthy();
+    });
```

Comment:
The existing regex you are modifying is not very robust and misses a lot of common use cases.

Examples of tests that it will either pass or fail that it should not would be for example:

```ts
    test('should verify valid wwww such as 4 of them.', () => {
      expect(isUrlInvalid('https://wwww.example.com')).toBeFalsy();
    });

    test('should validate characters such as %22 being part of a correct URL.', () => {
      expect(isUrlInvalid('https://www.exam%22ple.com')).toBeFalsy();
    });

    test('should validate characters incorrectly such as ;', () => {
      expect(isUrlInvalid('https://www.example.com[')).toBeTruthy();
    });
```

I think also this should include at least one test case for `http` such as:

```ts
    test('verifies valid http url', () => {
      expect(isUrlInvalid('http://www.elastic.co/')).toBeFalsy();
    });
```

If you look online you will see a lot of ambiguous language around the RFC
https://datatracker.ietf.org/doc/html/rfc3986 

But those I think are common use cases. The percents are very common in URL copying and pasting. I think that a better way would be to use `new URL` such as this below:

```ts
const allowedSchemes = ['http:', 'https:'];

export const isUrlInvalid = (url: string | null | undefined) => {
  try {
    if (url != null) {
      const urlParsed = new URL(url);
      if (allowedSchemes.includes(urlParsed.protocol)) {
        return false;
      }
    }
  } catch (error) {
    // intentionally left empty
  }
  return true;
};
```

Although this name is `isUrlInvalid` most developers prefer the positive case such as `isUrlValid` rather than the negative case.

Reply:
One more test case that would be helpful reading the code below is to check to ensure empty strings return false since there was an original `isEmpty(url)` and we want to ensure that the tests still pass after any refactoring. So this test would be helpful to add to preserve the original behavior first before the refactoring.

```ts
    test('verifies as valid when given an empty string', () => {
      expect(isUrlInvalid('')).toBeFalsy();
    });
```

A good piece of advice is to add all tests to exercise the current behaviors that we want to keep before refactoring. If you look at my re-factored code above I am introducing this behavioral change now because I am not checking this condition so it could end up as a bug. So in reality my code should be more something like this:

```ts
const allowedSchemes = ['http:', 'https:'];

export const isUrlInvalid = (url: string | null | undefined) => {
  try {
    if (url != null) {
      if (url === '') {
        return false;
      } else {
        const urlParsed = new URL(url);
        if (allowedSchemes.includes(urlParsed.protocol)) {
          return false;
        }
      }
    }
  } catch (error) {
    // intentionally left empty
  }
  return true;
};
```

It's a bit more noisy but it adheres to the original behavior which might involve allowing a completely empty string to work which I think the original author wanted.

One more test that would helpful to double check our understanding is what about empty spaces? If the original code did not allow spaces, then we should add that test case as well for ourselves and for other programmers later on such as:

```ts
    test('empty spaces should valid as not valid ', () => {
      expect(isUrlInvalid(' ')).toBeTruthy();
    });
```


A lot of times as I change and refactor I keep the "old code" commented out and then write the tests to ensure each behavior until I remove the older code for the newer refactored version. I like to swap between them as I encounter more of these types of things.